hmac_authentication/hmacauth.py

```diff
@@ -78,61 +86,60 @@ def sign_request(self, environ):
         return environ.get(self._signature_header)
 
-    @staticmethod
-    def _compare_signatures(header, computed):
-        if hmac.compare_digest(header.encode('utf8'), computed.encode('utf8')):
-            return HmacAuth.MATCH
-        else:
-            return HmacAuth.MISMATCH
-
-    def validate_request(self, environ):
-        '''Authenticates the request, returning the result code, the signature
-	from the header, and the locally-computed signature.
-    
-        Returns a tuple containing (the comparison result, the signature from
-        the request header, and the locally-computed signature). The
-        comparison result will be one of the following class constants:
-
-        - NO_SIGNATURE: the incoming result did not have a signature header
-        - INVALID_FORMAT: the signature header was not parseable
-        - UNSUPPORTED_ALGORITHM: the signature header specified an unsupported
-          algorithm
-        - MATCH: the signature from the request header matched the
-          locally-computed signature
-        - MISMATCH: the signature from the request header did not match the
-          locally-computed signature
-
-        These values can be passed to result_code_to_string() to return the
-        corresponding string value.
+    def authenticate_request(self, environ):
+        '''Authenticates the request by comparing HMAC signatures.
+        
+        Returns the result code, the signature from the header, and the
+        locally-computed signature as a AuthenticationResult.
         '''
         header = self.signature_from_header(environ)
         if header is None:
-            return HmacAuth.NO_SIGNATURE, None, None
+            return AuthenticationResult(
+                AuthenticationResultCodes.NO_SIGNATURE, None, None)
         components = header.split(' ')
 
         if len(components) != 2:
-            return HmacAuth.INVALID_FORMAT, header, None
+            return AuthenticationResult(
+                AuthenticationResultCodes.INVALID_FORMAT, header, None)
 
         digest_name = components[0]
         if not digest_name in hashlib.algorithms_available:
-            return HmacAuth.UNSUPPORTED_ALGORITHM, header, None
+            return AuthenticationResult(
+                AuthenticationResultCodes.UNSUPPORTED_ALGORITHM, header, None)
 
         computed = self._request_signature(
-            environ, getattr(hashlib, digest_name), digest_name)
-        return HmacAuth._compare_signatures(header, computed), header, computed
+            environ, getattr(hashlib, digest_name))
+        return AuthenticationResult(
+            _compare_signatures(header, computed), header, computed)
+
+
+class HmacMiddleware(object):
+    '''WSGI middleware for authenticating incoming HTTP requests via HmacAuth'''
+    
+    def __init__(self, app, hmac_auth):
+        self.app = app
+        self.hmac_auth = hmac_auth
+
+    def __call__(self, environ, start_response):
```

Comment:
This will raise errors because `abort` and `HTTPException` aren't defined. Both come from the very useful werkzeug library: `from werkzeug.exceptions import abort, HTTPException`. If we use this, we'll also want to add werkzeug to setup.py.